HTML API: Add table support

sirreal opened the pull request:
Trac ticket: Core-61576

Add support for table elements:
- TABLE
- THEAD, TBODY, TFOOT
- TR
- TD, TH
- COL, COLGROUP
- CAPTION

Not all necessary insertion modes are implemented in this PR, so e.g. "in caption" and "in select in table" modes are not implemented in this PR.

HTML5-lib test change (`./vendor/bin/phpunit --group html-api-html5lib-tests`):

```diff
-Tests: 610, Assertions: 275, Skipped: 335.
+Tests: 610, Assertions: 301, Skipped: 309.
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Work on table support

sirreal committed:
table processing

sirreal committed:
Disable FRAME / HEAD

sirreal committed:
In table body rules

sirreal committed:
prep step_in_row

sirreal committed:
Remove unsupported table element tests

sirreal committed:
phpcbf and implement in row

sirreal committed:
Add clear_up_to_last_marker to active formatting

This is an algorithm defined in the standard:
https://html.spec.whatwg.org/multipage/parsing.html#clear-the-list-of-active-formatting-elements-up-to-the-last-marker

sirreal committed:
Add step_in_cell method

sirreal committed:
Use class name for processor state constants access

sirreal committed:
Update since tags

sirreal committed:
Add close_cell method

sirreal committed:
Pop from open elements instead of removing items

sirreal committed:
Complete cases in step_in_table

sirreal committed:
Implement in_table_scope

sirreal committed:
Add HTML elements to has_element_in_scope handling

sirreal committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/add-table-support

sirreal committed:
Use insert_marker over set_marker

Use the method implemented in https://github.com/WordPress/wordpress-develop/pull/6982
to avoid duplicating the same functionality.

sirreal committed:
Use newly implemented step_in_X methods

Instead of bailing directly, call the appropriate step_in_ method.

sirreal committed:
Clean whitespace

sirreal committed:
Use bail method in case of foster parenting

sirreal committed:
Add mising cell insertion mode on enter td,th

sirreal committed:
PHPCBF

sirreal committed:
Use todo comments for parse errors

sirreal committed:
Remove EOF comment

sirreal committed:
Move stack methods to stack class

sirreal committed:
Add and use form_element pointer

sirreal committed:
Handle presumptuous tags as if they were comments

sirreal committed:
Add test for table > form > #comment

sirreal committed:
Pop FORM elements off the stack in tables

sirreal committed:
Be more consistent in parse error comments

sirreal commented:
@dmsnell This is ready for review.

sirreal committed:
Remove outdated "stub implementation" notes

sirreal committed:
Add return types

sirreal committed:
Handle whitespace in TABLE text

sirreal committed:
Fix table start tag handling

sirreal committed:
Remove "COL" from void tags test

The COL element can only appear in TABLE and is not suitable for the
void tag nesting test.

sirreal committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/add-table-support

sirreal committed:
Fix handling of table text according to specification

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1752,7 +1752,19 @@ private function step_in_table(): bool {
-				return $this->step_in_table_text();
+				$current_token = $this->bookmarks[ $this->state->current_token->bookmark_name ];
+				if (
+					strspn( $this->html, "\u{0000}\u{0009}\u{000A}\u{000C}\u{000D}\u{0020}", $current_token->start, $current_token->length ) === $current_token->length
```
note that there's no need for using the syntax here which has less support.

https://github.com/WordPress/wordpress-develop/pull/6977/files#diff-f822c7309a0a11d3993b378b85ef4a60588af0afac714f7e1590130aa4e300e5R1105

it's enough to use `\x00\x09\x0A\x0C\x0D\x20` which is I believe is identical to `\x00\t\r\f\n ` but the latter is far-more recognizable in code.

check out IN BODY as well where we take a different approach to this

https://github.com/WordPress/wordpress-develop/blob/339f70101549e7dd626495d6660f82872810a503/src/wp-includes/html-api/class-wp-html-processor.php#L1757-L1773

There's only a need to ignore the token if it's _only_ null bytes, otherwise we will rely on `get_modifiable_text()` to return the proper text.

Keep in mind that we should not be checking the raw HTML here. The spec is comparing already-decoded text, meaning that `&#x09;` is also ignored, since it decodes to U+09.

sirreal replied:
This comment is on a stale commit, although some if the review is still valid. I'll update the characters used.

sirreal replied:
> it's enough to use \x00\x09\x0A\x0C\x0D\x20 which is I believe is identical to \x00\t\r\f\n  but the latter is far-more recognizable in code.

I actually find the hex form easier to work with because I'm going back and forth with specifications that has definitions like this:

> ASCII whitespace is U+0009 TAB, U+000A LF, U+000C FF, U+000D CR, or U+0020 SPACE.

I'm not confident about the special escape sequences for all of these characters from memory.

How would you feel about adding a constant to one of the classes that can include documentation and can be reused in multiple places?

```php
/**
 * A string of the ASCII whitespace characters according to the Infra Standards.
 * 
 * > ASCII whitespace is U+0009 TAB, U+000A LF, U+000C FF, U+000D CR, or U+0020 SPACE.
 * @see https://infra.spec.whatwg.org/#ascii-whitespace
 * 
 * @var string
 */
const ASCII_WHITESPACE = /* I no longer care about this representation */;
```

dmsnell replied:
I'm reluctant to make sweeping changes at the moment. One problem with the constant is that many places in the spec add additional characters, so now on top of seeing the codes, you have to also correlate them with the additions, and if the additions are already contained within the constant.

If we feel strongly enough to use `\x09` instead of `\t` then we can address that in a comprehensive change. As it stands, everything uses the standard codes and I would prefer not introducing arbitrary change here.

Would be good to poll people and see if they recognize `"\x09\x0a\x0c\x0d\x20"` easier or `"\t\r\f\n "`

sirreal replied:
I've changed the whitespace to use the more common escapes and updated the existing place to follow the ordering of whitespace characters in the specification. I also quoted that part of the spec in this case so there should be a clear correspondence and no confusion.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1752,7 +1752,19 @@ private function step_in_table(): bool {
+					$this->insert_html_element( $this->state->current_token );
+					return true;
+				}
+				$this->bail( 'Foster parenting is not supported.' );
```
can we just say that this parser doesn't support foster parenting and be good with it? I don't have as good of a bearing on this as I do with SCRIPT support, where it's clear that the server-side parser can't support it.

foster parenting impacts structure and CSS selectors. what would happen if implemented the rules but foster parenting were always disabled? 🤔

sirreal replied:
[To be fully compliant, I don't think that is allowed](https://html.spec.whatwg.org/#creating-and-inserting-nodes):

> While the parser is processing a token, it can enable or disable foster parenting. This affects the following algorithm.
> *Note* Foster parenting happens when content is misnested in tables.

Are you concerned about proceeding with this change while foster parenting is not supported, or is the question about whether we need to support it in the long term?

sirreal replied:
> foster parenting impacts structure and CSS selectors. what would happen if implemented the rules but foster parenting were always disabled?

If it continued to bail, then we'd refuse to process HTML with misnested content in tables. This would be a limitation and a deviation from the specification, where HTML with a valid representation would be rejected by the HTML Processor.

In order to support foster parenting, I think we'd need a design shift that would likely be harmful to general performance and memory overhead in certain circumstances. When reaching a table, I believe we'd need to _process the entire table_ before pausing on the table node because of cases like this:

```html
<table>
  <!-- A gigantic valid table -->
  <span>Oops! In HTML this appears at the very end is actually before the table!</span>
</table>
```

If we expect to reach those nodes in order, the `SPAN` is the sibling of the `TABLE` immediately preceding it, so `SPAN` should be reached first.

If we ignore ordering, then matching CSS selectors (depending on the supported selectors) we could recognize that the `SPAN` is not a descendant of the TABLE, but of the TABLE's parent.

dmsnell replied:
> Are you concerned about proceeding with this change while foster parenting is not supported, or is the question about whether we need to support it in the long term?

Long term - not about this PR

dmsnell replied:
> When reaching a table, I believe we'd need to process the entire table before pausing on the table node because of cases like this:

I agree, which is why I've always wanted it to be opt-in.

But on the other hand, what if we just say _this is not a foster-parenting-enabled parser_ and note that as a limitation? That may not be a divergence from the spec because we have chosen to build a parser without foster parenting enabled.

sirreal replied:
> what if we just say this is not a foster-parenting-enabled parser and note that as a limitation

That's certainly viable in the short term, the HTML Processor does not support everything now. We don't need to finalize the decision of whether to support fostering right now, it can be deferred until later.

There may also be a good middle ground with budgeting where the processor could try to parse the table up to a certain point, then bail after a threshold is reached.

**For now** I think we should certainly note that fostering is not enabled (and explain what that means). If and when it makes sense, we can add support later.

dmsnell replied:
well I think you are answering a different question than I asked, but I believe that I found my answer in the spec and now I remember getting to this point before.

the question is whether enabling foster parenting is a choice of the parser or not. apparently it's not, and the toggling of foster parenting is triggered in the [anything else](https://html.spec.whatwg.org/#parsing-main-intable:foster-parent) rule of `IN TABLE` insertion mode.

I keep hoping this is more like the scripting tag, or the speculative parser, or whether we report errors and the decision to allow foster parenting is a choice of the parser. alas. we'll just carry on as usual.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1723,8 +1742,207 @@ private function step_in_body() {
+						return $this->step();
+					}
+
+					// Whitespace-only text nodes are inserted in-place.
+					if ( strlen( $text ) === strspn( $text, "\u{0009}\u{000A}\u{000C}\u{000D}\u{0020}", 0, strlen( $text ) ) ) {
```
```suggestion
					/*
					 * This follows the rules for "in table text" insertion mode.
					 *
					 * Whitespace-only text nodes are inserted in-place. Otherwise
					 * foster parenting is enabled and the nodes would be
					 * inserted out-of-place.
					 *
					 * > If any of the tokens in the pending table character tokens
                                         * > list are character tokens that are not ASCII whitespace,
                                         * > then this is a parse error: reprocess the character tokens
                                         * > in the pending table character tokens list using the rules
                                         * > given in the "anything else" entry in the "in table"
                                         * > insertion mode.
                                         * >
                                         * > Otherwise, insert the characters given by the pending table
                                         * > character tokens list.
					 *
					 * @see https://html.spec.whatwg.org/#parsing-main-intabletext
                                         */
					if ( strlen( $text ) === strspn( $text, "\x09\x0A\x0C\x0D\x20", 0, strlen( $text ) ) ) {
```

sirreal committed:
Expand text processing comment and whitespace special character form

sirreal committed:
Fix comment whitespace

sirreal committed:
Clarify empty check after processing and null-remove

sirreal committed:
Use consistent "\n" style character escapes

Follow the most common whitespace character escape forms.
Quote the spec referencing the ASCII whitespace characters.
List the characters in the same order mentioned in the spec.

dmsnell committed:
Merge branch 'trunk' into html-api/add-table-support

dmsnell commented:
@sirreal I've merged `trunk` into this branch, and resolved conflicts. Please review and ensure I didn't accidentally break your changes.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2127,229 @@ private function step_in_body(): bool {
+					$this->state->stack_of_open_elements->current_node_is( 'THEAD' ) ||
+					$this->state->stack_of_open_elements->current_node_is( 'TR' )
+				) {
+					$text = str_replace( "\0", '', $this->get_modifiable_text() );
```
this should probably be removed given that `get_modifiable_text()` is responsible for replacing NULL bytes.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2127,229 @@ private function step_in_body(): bool {
+					$this->state->stack_of_open_elements->current_node_is( 'TEMPLATE' ) ||
+					$this->state->stack_of_open_elements->current_node_is( 'TFOOT' ) ||
+					$this->state->stack_of_open_elements->current_node_is( 'THEAD' ) ||
+					$this->state->stack_of_open_elements->current_node_is( 'TR' )
```
maybe this helps, but in the case of comparing against multiple items I don't find the call to `current_node_is()` more helpful can capturing the current node name and comparing to the names.

sirreal replied:
I'll apply this suggestion, I don't feel strongly either way.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2127,229 @@ private function step_in_body(): bool {
+			 */
+			case '+COL':
+				$this->state->stack_of_open_elements->clear_to_table_context();
+				$this->insert_html_element(
```
worth a comment

```php
/*
 * > Insert an HTML element for a "colgroup" start tag token with no attributes,
 * > then switch the insertion mode to "in column group".
 */
```

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2127,229 @@ private function step_in_body(): bool {
+			case '+TH':
+			case '+TR':
+				$this->state->stack_of_open_elements->clear_to_table_context();
+				$this->insert_html_element(
```
same here: worth a comment

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2127,229 @@ private function step_in_body(): bool {
+			case '+INPUT':
+				$type_attribute = $this->get_attribute( 'type' );
+				if ( ! is_string( $type_attribute ) || 'hidden' !== strtolower( $type_attribute ) ) {
+					break;
```
need a comment indicating that this should fall to "anything else" below
perhaps it would be good to add some `goto`s in this and other modes where we have this jump

sirreal replied:
I've used goto, that seems sufficient. The comment quoting the spec above includes the instructions to go to the anything else case.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2203,7 +2425,93 @@ private function step_in_column_group(): bool {
+				$this->state->stack_of_open_elements->clear_to_table_body_context();
+				$this->insert_html_element(
+					new WP_HTML_Token( null, 'TR', false )
+				);
```
we could probably benefit from something like this

```php
$this->create_virtual_node( 'TR' );
```

what do you think?

sirreal replied:
That would be helpful for some of these generated tags in tables, although it only happens in a few places and they're already covered. I'd like to consider it for a follow-up but not include it in this PR.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2241,7 +2636,85 @@ private function step_in_row(): bool {
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TD' ) &&
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TH' )
+				) {
+					throw new Exception( 'Assertion failed @todo better message' );
```
earmark: we shouldn't leave the exception in here

sirreal replied:
This is an assertion from the specification. Should we bail? Should we have an assert method?

If this assertion fails, it means something has gone very wrong and it's not clear what we can do other than fail.

dmsnell replied:
tell me what you think.
of course something has gone wrong, but I also don't want to throw an `Exception()`

how likely is it that we'd get here? is it dependent on the document, or on our code?
what are the repercussions of going forward past this assertion if it fails?

sirreal replied:
Here are the relevant specs:

> > [Assert](https://infra.spec.whatwg.org/#assert): The [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements) [has a td or th element in table scope](https://html.spec.whatwg.org/multipage/parsing.html#has-an-element-in-table-scope).
> …
> [3.10. Assertions](https://infra.spec.whatwg.org/#assertions)
> To improve readability, it can sometimes help to add assertions to algorithms, stating invariants. To do this, write "Assert:", followed by a statement that must be true. If the statement ends up being false that indicates an issue with the document using the Infra Standard that should be reported and addressed.
> …
> Note: Since the statement can only ever be true, it has no implications for implementations.

---

> how likely is it that we'd get here?

I think this is completely dependent on whether there are bugs in the implementation. I'm assuming that the spec is correct and this assertion should hold, which only leaves the implementation.

> is it dependent on the document, or on our code?

It would depend on both, but the problem would be in our code. We'd need a document with the right conditions to trigger a failure path.

> what are the repercussions of going forward past this assertion if it fails?

We'd proceed to follow these steps:

> [Generate implied end tags](https://html.spec.whatwg.org/multipage/parsing.html#generate-implied-end-tags).
> 
> If the [current node](https://html.spec.whatwg.org/multipage/parsing.html#current-node) is not now a [td](https://html.spec.whatwg.org/multipage/tables.html#the-td-element) element or a [th](https://html.spec.whatwg.org/multipage/tables.html#the-th-element) element, then this is a [parse error](https://html.spec.whatwg.org/multipage/parsing.html#parse-errors).
> 
> Pop elements from the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements) until a [td](https://html.spec.whatwg.org/multipage/tables.html#the-td-element) element or a [th](https://html.spec.whatwg.org/multipage/tables.html#the-th-element) element has been popped from the stack.
> 
> [Clear the list of active formatting elements up to the last marker](https://html.spec.whatwg.org/multipage/parsing.html#clear-the-list-of-active-formatting-elements-up-to-the-last-marker).
> 
> Switch the [insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#insertion-mode) to "[in row](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-intr)".

So we'd be calling `pop_until` with a tag we'd never reach (or maybe we'd reach a cell in an ancestor table). We'd either pop until the wrong table or I supposed pop until that failed or maybe the stack is empty. It would depend on the exact behavior of `pop_until`.

It's not clear exactly what would happen, but things would be a complete mess.

---

I'm not fully decided. If we hit this condition, the implementation has a bad bug and the processor gets into a bad state with undefined behavior. I'd really like to know if we have a bug like that.

On the other hand, implementations are also free to ignore the assertion because it should always be true. If we trust the implementation is correct, we can omit the assertion completely.

🤔 Let's remove the assertion.

dmsnell replied:
thanks for gathering all this. I think if we want to assert this we should do so from a test or through running tests to parse documents in the wild. should we encounter this situation on a production site, there's apparently nothing anyone can do about it, so I think closing too many elements is not the worst failover for a condition that should never trigger

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessor.php:
```diff
@@ -331,7 +331,6 @@ public static function data_void_tags() {
 			'AREA'   => array( 'AREA' ),
 			'BASE'   => array( 'BASE' ),
 			'BR'     => array( 'BR' ),
-			'COL'    => array( 'COL' ),
```
was this a mistake? I see no reason to treat `COL` as if it's not a void tag.

sirreal replied:
No, this tag didn't work in the void tag tests because it's only present in tables. That was fixed as part of #6972 so I'll revert the change here.

dmsnell replied:
tests have been tricky at places, because we no longer have the concept of being able to step into _any_ tag, since stepping into that tag may ignore it or create other tags along the way 🙃

dmsnell reviewed (commented):
this one's ready for real review and bound for merging!

sirreal committed:
Remove redundant null byte text replacement

sirreal committed:
Apply suggestion to compare multiple elements against node name

sirreal committed:
Add spec quote when generating a COLGROUP token

The spec inserts some elements not present in the DOM to preserve proper
table structure. Add explanatory comment.

sirreal committed:
Add spec quote when generating a TBODY token

The spec inserts some elements not present in the DOM to preserve proper
table structure. Add explanatory comment.

sirreal committed:
Use goto for safer move to "anything else" condition

sirreal committed:
Revert "Remove "COL" from void tags test"

This reverts commit d14eaf318a3e51428b41cb8c8dee179ce6082d15.

sirreal committed:
fixup! Apply suggestion to compare multiple elements against node name

sirreal committed:
Add comment for no-quirks p table nesting

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2127,7 +2131,238 @@ private function step_in_body(): bool {
+					 *
+					 * @see https://html.spec.whatwg.org/#parsing-main-intabletext
+					 */
+					if ( strlen( $text ) === strspn( $text, "\t\n\f\r ", 0, strlen( $text ) ) ) {
```
what is the point of supplying the default parameters to `strspn()`? (I know that the default for `$length` is `NULL`, but the behavior of that is `strlen($text)`)

sirreal replied:
```suggestion
					if ( strlen( $text ) === strspn( $text, "\t\n\f\r " ) ) {
```

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2203,7 +2438,93 @@ private function step_in_column_group(): bool {
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'THEAD' ) &&
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TFOOT' )
+				) {
+					// @todo Indicate a parse error once it's possible.
```
just curious about this, because these elements are implicitly created on `+TABLE`, so I think the only case we could get here is with a TEMPLATE

```html
<table>
<td><template>This </table> does not close the outer one</template>
</table>
```

sirreal replied:
This is just following the spec (quoted below). I'm reluctant to deviate from the spec in all but the most obvious places. There are many places the switch modes and reprocess tokens _or_ process tokens following the rules for other modes.

> A character token, if the [current node](https://html.spec.whatwg.org/multipage/parsing.html#current-node) is [table](https://html.spec.whatwg.org/multipage/tables.html#the-table-element), [tbody](https://html.spec.whatwg.org/multipage/tables.html#the-tbody-element), [template](https://html.spec.whatwg.org/multipage/scripting.html#the-template-element), [tfoot](https://html.spec.whatwg.org/multipage/tables.html#the-tfoot-element), [thead](https://html.spec.whatwg.org/multipage/tables.html#the-thead-element), or [tr](https://html.spec.whatwg.org/multipage/tables.html#the-tr-element) element
> Let the pending table character tokens be an empty list of tokens.
> 
> Let the [original insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#original-insertion-mode) be the current [insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#insertion-mode).
> 
> Switch the [insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#insertion-mode) to "[in table text](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-intabletext)" and reprocess the token.

dmsnell replied:
there was no hint of asking you to deviate from the spec; all I was doing was observing an interesting oddity since the only elements we're checking are ones that should always be open. I think that `TEMPLATE` is the only exception. this was just exploring the rules.

if this turns out to be true, it might be possible one day to verify and take advantage of. for now, all it did was seem odd when I saw it, and then I realized there's an escape hatch in almost all of these situations - the `TEMPLATE` element.

sirreal committed:
Remove strspn default args

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2241,7 +2649,85 @@ private function step_in_row(): bool {
+			case '+TH':
+			case '+THEAD':
+			case '+TR':
+				// Assert: The stack of open elements has a td or th element in table scope.
+				if (
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TD' ) &&
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TH' )
+				) {
+					throw new Exception( 'Assertion failed @todo better message' );
+				}
+
```
Remove the assertion. It's invariant and optional according to spec.
```suggestion
```

sirreal committed:
Remove assertion in implementation from HTML spec

Assertions from the spec are explicitly optional in implementation.

Remove the assertion, it's checking an invariant condition.

dmsnell committed:
Adjust code after review.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -765,6 +765,7 @@ public function expects_closer( $node = null ): ?bool {
 		}
 
 		return ! (
+			( $node->has_self_closing_flag ?? false ) ||
```
Interestingly, @sirreal, this is no longer required, since we moved to processing the event queue. I'm leaving it in for now, though, while pondering the later change which avoids immediately popping the FORM off the stack of open elements.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2179,12 +2180,10 @@ private function step_in_table(): bool {
 					 * @see https://html.spec.whatwg.org/#parsing-main-intabletext
 					 */
-					if ( strlen( $text ) === strspn( $text, "\t\n\f\r " ) ) {
+					if ( strlen( $text ) === strspn( $text, " \t\f\r\n" ) ) {
```
Updated these @sirreal purely to harmonize with the existing checks, which I think are universally in this order, plus whitespace is defined in the spec and used somewhat ubiquitously around here. I don't think it's necessary to repeat here.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2208,7 +2207,7 @@ private function step_in_table(): bool {
 			 */
 			case 'html':
-				// @todo Indicate a parse error once it's possible.
+				// Parse error: ignore the token.
```
Your critique of these changes is probably sound. I'm trying to harmonize existing comments and `@todo` statements. I _think_ that in cases of unexpected tokens like this I'm leaving it as `Parse error: ignore the token` mostly to give intention behind why we're returning `$this->step()`. In cases where we have another violation, such as an element not being in scope, I'm leaving the `@todo`. Perhaps we remove them all during 6.7 🤷‍♂️

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2235,11 +2234,12 @@ private function step_in_table(): bool {
 				 * > then switch the insertion mode to "in column group".
 				 */
-				$this->insert_html_element( new WP_HTML_Token( null, 'COLGROUP', false ) );
+				$this->insert_virtual_node( 'COLGROUP' );
```
When I got into this @sirreal I preferred giving a bookmark to the virtual nodes, letting people return to them. Once I did that there was a reasonable amount of ceremony around creating and inserting the token, so I went ahead and created this helper method. Please let me know if you object to it.

sirreal replied:
I like the new method 👍

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2347,10 +2354,12 @@ private function step_in_table(): bool {
 				}
+
+				// This FORM is special because it immediately closes and cannot have other children.
+				$this->state->current_token->has_self_closing_flag = true;
```
This is where I'm pondering.

sirreal replied:
What's the motivation for doing it this way instead of adding and popping right here?

dmsnell committed:
Remove typo.

dmsnell reviewed (commented):
@sirreal I fixed two logic issues:

 - in `IN CELL` we have to check if the current node has the same name as the matched token, but we were only checking that it had a name at all.
 - in `IN TABLE BODY` for `-TBODY, …` we needed to clear to table body context.

After our short discussion about it I went ahead with `insert_virtual_node()` because I want that to have a bookmark, and I think it might be important later to add things like namespace and whether it's virtual.

I've made comment and other minor changes, which can all be reverted if you think it's best as it was before.

sirreal committed:
Fix HTML spec quoting close_cell method

The "> …" quoting omitted part of the quoted spec. Include the complete
spec in the quoting.

sirreal commented:
There's an error in the new form test with the change to auto-close the form where the form closer is not visited by `next_token`. It's fixed with this patch, although I haven't investigated why:

```diff
diff --git before/src/wp-includes/html-api/class-wp-html-processor.php after/src/wp-includes/html-api/class-wp-html-processor.php
index 1e6a18389b..0d1c7c1496 100644
--- before/src/wp-includes/html-api/class-wp-html-processor.php
+++ after/src/wp-includes/html-api/class-wp-html-processor.php
@@ -2356,10 +2356,9 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 				}
 
 				// This FORM is special because it immediately closes and cannot have other children.
-				$this->state->current_token->has_self_closing_flag = true;
-
 				$this->insert_html_element( $this->state->current_token );
 				$this->state->form_element = $this->state->current_token;
+				$this->state->stack_of_open_elements->pop();
 				return true;
 		}
 
```

I'll push that change and tests should be passing again. It's effectively reverting part of your updates so happy to discuss further.

sirreal committed:
Use pop instruction for form elements that are immediately popped in table

sirreal commented:
@dmsnell The changes you've pushed look good, thanks for catching a couple of bugs and I do like the method for creating virtual nodes.

The last remaining thing is to sort out the FORM element in table insertion mode that's immediately popped.

dmsnell committed:
Remove unwanted change to expects-closer

dmsnell commented:
> I'll push that change and tests should be passing again. It's effectively reverting part of your updates so happy to discuss further.

I was considering this because I wanted the code to function as the void tags function, but I can't figure out what's wrong. Along the way I found some bugs in `next_token()` I introduced when fixing breadcrumbs, but the fix is incoming and easy. We don't close out the remaining open stack elements at the end of the document _because they never get popped_.

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/add-table-support

dmsnell commented:
Merged in [[58806]](https://core.trac.wordpress.org/changeset/58806)
https://github.com/wordpress/wordpress-develop/commit/5351cd20f3d5eb890cb7b53270d01b626ce02f7d
